src/drivers/QtTestDriver.cpp

```diff
@@ -0,0 +1,203 @@
+        m_oldStdErr = secure_dup(STD_ERR_FD);
+        secure_dup2(m_pipe[WRITE], STD_OUT_FD);
+        secure_dup2(m_pipe[WRITE], STD_ERR_FD);
+        m_capturing = true;
+#ifndef WINDOWS
+        secure_close(m_pipe[WRITE]);
+#endif
+    }
+
+    bool IsCapturing() {
+        QMutexLocker locker(&m_mutex);
+        return m_capturing;
+    }
+
+    void EndCapture() {
+        QMutexLocker locker(&m_mutex);
+        if (!m_capturing)
+            return;
+
+        m_captured.clear();
+        secure_dup2(m_oldStdOut, STD_OUT_FD);
+        secure_dup2(m_oldStdErr, STD_ERR_FD);
+
+        const int bufSize = 1025;
+        char buf[bufSize];
+        int bytesRead = 0;
+        bool fd_blocked(false);
+        do {
+            bytesRead = 0;
+            fd_blocked = false;
+#ifdef WINDOWS
+            if (!eof(m_pipe[READ]))
+                bytesRead = read(m_pipe[READ], buf, bufSize-1);
+#else
+            bytesRead = read(m_pipe[READ], buf, bufSize-1);
+#endif
+            if (bytesRead > 0) {
+                buf[bytesRead] = 0;
+                m_captured += buf;
+            } else if (bytesRead < 0) {
+                fd_blocked = (errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR);
+                if (fd_blocked)
+                    QThread::msleep(10);
+            }
+        }
+        while(fd_blocked || bytesRead == (bufSize-1));
+
+        secure_close(m_oldStdOut);
+        secure_close(m_oldStdErr);
+        secure_close(m_pipe[READ]);
+#ifdef WINDOWS
+        secure_close(m_pipe[WRITE]);
+#endif
+        m_capturing = false;
+    }
+
+    std::string GetCapture() {
+        QMutexLocker locker(&m_mutex);
+        return m_captured;
+    }
+
+private:
+    enum PIPES { READ, WRITE };
+
+    int secure_dup(int src) {
+        int ret = -1;
+        bool fd_blocked = false;
+        do {
+             ret = dup(src);
+             fd_blocked = (errno == EINTR ||  errno == EBUSY);
+             if(fd_blocked)
+                 QThread::msleep(10);
+        } while (ret < 0);
+        return ret;
+    }
+
+    void secure_pipe(int * pipes) {
+        int ret = -1;
+        bool fd_blocked = false;
+        do {
+#ifdef WINDOWS
+            ret = pipe(pipes, 65536, O_BINARY);
+#else
+            ret = pipe(pipes) == -1;
+#endif
+            fd_blocked = (errno == EINTR ||  errno == EBUSY);
+            if (fd_blocked)
+                QThread::msleep(10);
+        } while (ret < 0);
+    }
+
+    void secure_dup2(int src, int dest) {
+        int ret = -1;
+        bool fd_blocked = false;
+        do {
+             ret = dup2(src,dest);
+             fd_blocked = (errno == EINTR ||  errno == EBUSY);
+             if (fd_blocked)
+                QThread::msleep(10);
+        } while (ret < 0);
+    }
+
+    void secure_close(int & fd) {
+        int ret = -1;
+        bool fd_blocked = false;
+        do {
+             ret = close(fd);
+             fd_blocked = (errno == EINTR);
+             if (fd_blocked)
```

Comment:
`EINTR` indicates that the syscall was interrupted by a syscall and can be retried immediately. There's no point in waiting first. (The same applies to other syscalls where `EINTR` is returned).